### ISSUE TITLE
False warning about unused constant

### ISSUE THREAD

**Issue by Flying-Toast** [rust-lang/rust#73637]
If an unused associated function contains a constant, rustc warns that it is unused (even if it is used within the function).

This:
```Rust
struct Five;

impl Five {
    fn return_five() -> i32 {
        const FIVE: i32 = 5;
        FIVE
    }
}
```
emits the warning ``warning: constant item is never used: `FIVE` ``.

---

The bug does not occur for constants in top level functions,
```Rust
fn return_five() -> i32 {
    const FIVE: i32 = 5;
    FIVE
}
```
or for *variables* in associated functions (or top level functions):
```Rust
struct Five;

impl Five {
    fn return_five() -> i32 {
        let five = 5;
        five
    }
}
```

**Comment by RalfJung** [comment]
This is an instance of a more general problem (probably already tracked elsewhere) where if an unused function `foo` is the only place some other function `bar` is called, then rustc warns about both being unused. One can disagree about whether this is the right behavior or not, but AFAIK it is at least consistent across functions, constants etc.
